modules/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,433 @@
+    /* Reset line */
+    OneWire_Reset(OneWireStruct);
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWireStruct, ROM);
+    /* Start temperature conversion */
+    OneWire_WriteByte(OneWireStruct, DS18B20_CMD_CONVERTTEMP);
+    
+    return 1;
+}
+
+char DS18B20_StartAll(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return DS18B20_USAGE_ERROR;
+    /* Reset pulse */
+    OneWire_Reset(OneWireStruct);
+    /* Skip rom */
+    OneWire_WriteByte(OneWireStruct, ONEWIRE_CMD_SKIPROM);
+    /* Start conversion on all connected devices */
+    OneWire_WriteByte(OneWireStruct, DS18B20_CMD_CONVERTTEMP);
+    return 1;
+}
+
+char DS18B20_Read(OneWire_t* OneWireStruct, uint8_t *ROM, float *destination) {
+    if (!OneWireStruct)   return DS18B20_USAGE_ERROR;
+    if (!ROM)             return DS18B20_USAGE_ERROR;
+    if (!destination)     return DS18B20_USAGE_ERROR;
+    if (!DS18B20_Is(ROM)) return DEVICE_NOT_DS18B20;
+
+    uint16_t temperature;
+    unsigned char resolution;
+    int8_t digit, minus = 0;
+    float decimal;
+    unsigned char i = 0;
+    unsigned char data[DS18B20_DATA_LEN];
+    unsigned char crc;
+    
+    /* Check if device is DS18B20 */
+    
+    
+    /* Check if line is released, if it is, then conversion is complete */
+    if (OneWire_ReadBit(OneWireStruct)==0) {
+        /* Conversion is not finished yet */
+        return DS18B20_CONVERSION_IN_PROGRESS; 
+    }
+
+    /* Reset line */
+    OneWire_Reset(OneWireStruct);
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWireStruct, ROM);
+    /* Read scratchpad command by OneWireStruct protocol */
+    OneWire_WriteByte(OneWireStruct, ONEWIRE_CMD_RSCRATCHPAD);
+    
+    /* Get data */
+    for (i = 0; i < DS18B20_DATA_LEN; i++) {
+        /* Read byte by byte */
+        data[i] = OneWire_ReadByte(OneWireStruct);
+    }
+    
+    /* Calculate CRC */
+    crc = OneWire_CRC8(data, DS18B20_DATA_CRC_BYTE);
+    
+    /* Check if CRC is ok */
+    if (crc != data[DS18B20_DATA_CRC_BYTE]) {
+        /* CRC invalid */
+        return DS18B20_CRC_INVALID;
+    }
+    
+    /* First two bytes of scratchpad are temperature values */
+    temperature = data[DS18B20_DATA_LSB] | (data[DS18B20_DATA_MSB] << 8);
+
+    /* Reset line */
+    OneWire_Reset(OneWireStruct);
+    
+    /* Check if temperature is negative */
+    if (temperature & DS18B20_TEMP_SIGN_BITMASK) {
+        /* Two's complement, temperature is negative */
+        temperature = ~temperature + 1;
+        minus = 1;
+    }
+
+    /* Get sensor resolution */
+    /* Extract resolution bits with &, Ignore unused bits in the configuration register 
+    which are reserved for internal use by the device. Add 9 to convert R0 and R1 -> 9-12 for 
+    bits in resolution */
+    resolution = ((data[DS18B20_CONFIG_REGISTER_BYTE] & DS18B20_CONFIG_REGISTER_R0_R1_BITMASK) 
+        >> DS18B20_CONFIG_REGISTER_RESERVED_BITS) + 9;
+    
+    /* Store temperature integer digits and decimal digits */

```

Comment:
lines 118 to 115 might be best to be refactored into a function that builds the float from the data.  Would make this function easier to follow and read.